File: Python_Project/src/python_project/Student.py
from Transcript import Transcript
from Grade import Grade


class Student:
    # Computed attributes may be declared as properties in methods
    # Semester, Instructor (nonprimitive types might be declared as properties)

    def __init__(self, student_id, name, surname, semester, advisor=None):
        # We can convert all course lists(Active, Past, Failed)
        # To map so that we can hold grades or keep going with 1 grade map as in JAVA
        self.name = name
        self.surname = surname
        self.student_id = student_id
        self.advisor = advisor
        self.active_courses = list()
        self.course_basket = list()
        self.past_courses = list()
        self.non_taken_courses = list()
        self.failed_courses = list()
        self.semester = semester
        self.gpa = 0
        self.cgpa = 0
        self.completed_credit = 0
        self.grade_map = dict()
        self.cumulative_credit = 0
        self.active_credit = 0
        self.cumulative_grade = 0
        self.active_grade = 0
        self.transcript = Transcript(self)

    def next_semester(self, semester):

        self.create_grades()
        self.calculate_gpa()
        self.calculate_cgpa()
        self.transcript.add_gpa(self.gpa, self.active_credit, self.cgpa, self.cumulative_credit)
        self.active_grade = 0
        self.active_credit = 0
        self.semester = semester
        self.transcript.create_semester()
        for course in self.active_courses:
            course.students.remove(self)
        self.active_courses.clear()

    def create_grades(self):
        for course in self.active_courses:
            grade = Grade()
            self.grade_map[course] = grade
            if grade.letter_grade == "FF":
                self.failed_courses.append(course)
            else:
                self.completed_credit += course.credit
                self.past_courses.append(course)
            self.active_grade += course.credit * grade.grade
            self.transcript.add_course(course.course_id, grade.letter_grade)
        self.cumulative_grade += self.active_grade

    def calculate_gpa(self):
        self.gpa = self.active_grade / self.active_credit

    def calculate_cgpa(self):
        self.cgpa = self.cumulative_grade / self.cumulative_credit

    def basket_to_active_course(self):
        """create grade
        We can assign next semester here but it is not logical
        clear basket"""
        self.active_courses.extend(self.course_basket)
        for course in self.course_basket:

            if course in self.failed_courses:
                self.cumulative_credit -= course.credit
                self.failed_courses.remove(course)
            self.active_credit += course.credit
            self.cumulative_credit += course.credit
            course.students.append(self)
        self.course_basket.clear()

    def add_error(self, error, course):
        self.non_taken_courses.append(course)
        self.transcript.add_error(error)

    def add_course_basket(self, course):
        self.course_basket.append(course)

    def remove_from_basket(self, will_be_removed):
        self.course_basket = list(set(self.course_basket) - set(will_be_removed))

File: Java_Project/src/main/java/Student.java
import org.apache.log4j.Logger;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Student extends Person {//A class for Students and it extends from Person class
    private Semester semester;//semester attribute to take semester of students that passed or active
    private Map<Course, Grade> gradeMap = new HashMap<>();//gradeMap is a HashMap for grading
    private float gpa = 0;//attribute for gpa
    private float cgpa = 0;//attribute for cgpa
    private Instructor advisor;//advisor attribute for instructor of the student
    private int completedCredit;//completedCredit takes the credits that student has completed
    private List<Course> activeCourses = new ArrayList<>();//ArrayList takes the active courses of the student
    private List<Course> courseBasket = new ArrayList<>();//ArrayList for student's course basket
    private List<Course> pastCourses = new ArrayList<>();//ArrayList takes the student's past courses
    private List<Course> nonTakenCourses = new ArrayList<>();//ArrayList takes the courses that student has not taken yet
    private List<Course> failedCourses = new ArrayList<>();// ArrayList takes the student's failed courses
    private List<Error> errors = new ArrayList<>();//ArrayList for errors that occurred
    private Transcript transcript;//transcript attribute for student's transcript

    public Student() {
        super();
    }

    public Student(int id, String name, String surname, List<String> emails, Semester semester) {
        super(id, name, surname, emails);
        this.semester = semester;
    }

    public Map<Course, Grade> getGradeMap() {
        return gradeMap;
    }

    //This can be achieved by getters and setters at StudentExpert
    public Semester getSemester() {
        return semester;
    }


    public void setSemester(Semester semester) {
        this.semester = semester;
    }

    public float getGpa() {
        return gpa;
    }

    public void setGpa(float gpa) {
        this.gpa = gpa;
    }

    public float getCgpa() {
        return cgpa;
    }

    public void setCgpa(float cgpa) {
        this.cgpa = cgpa;
    }

    public Instructor getAdvisor() {
        return advisor;
    }

    public void setAdvisor(Instructor advisor) {
        this.advisor = advisor;
    }

    public int getCompletedCredit() {
        return completedCredit;
    }

    public void setCompletedCredit(int completedCredit) {
        this.completedCredit = completedCredit;
    }

    public void clearActiveCourses() {
        this.activeCourses.clear();
    }

    public void basketToActiveCourses() {
        this.activeCourses.addAll(courseBasket);
    }

    public void clearCourseBasket() {
        this.courseBasket.clear();
    }

    public void addActiveCredit() {

        for (Course course : courseBasket) {
            if (failedCourses.contains(course)) {
                double cumulativeCredit = transcript.getCumulativeCredit();
                cumulativeCredit -= course.getCredit();
                transcript.setCumulativeCredit(cumulativeCredit);
                failedCourses.remove(course);
            }

            double activeCredit = transcript.getActiveCredit();
            activeCredit += course.getCredit();
            transcript.setActiveCredit(activeCredit);

            double cumulativeCredit = transcript.getCumulativeCredit();
            cumulativeCredit += course.getCredit();
            transcript.setCumulativeCredit(cumulativeCredit);
            course.addStudentToArraylist(this);
        }
    }

    public List<Course> getCourseBasket() {
        return courseBasket;
    }


    public void addCourseToBasket(Course course) {
        this.courseBasket.add(course);
    }

    public List<Course> getPastCourses() {
        return pastCourses;
    }


    public void addPastCourse(Course course) {
        this.pastCourses.add(course);
    }

    public List<Course> getNonTakenCourses() {
        return nonTakenCourses;
    }

    public void addNonTakenCourse(Course course) {
        this.nonTakenCourses.add(course);
    }

    public List<Course> getFailedCourses() {
        return failedCourses;
    }

    public void addFailedCourse(Course course) {
        this.failedCourses.add(course);
    }

    public List<Error> getErrors() {
        return errors;
    }

    public void setErrors(List<Error> errors) {
        this.errors = errors;
    }

    public void addError(Error error) {
        this.errors.add(error);
        transcript.addError(error);
        Logger logger = Logger.getLogger(this.getClass().getName());
        logger.info(error.raiseError());
    }

    public void calculateGPA() {
        double activeCredit = transcript.getActiveCredit();
        double activeGrade = transcript.getActiveGrade();
        double result = activeGrade / activeCredit;
        setGpa((float) result);
    }

    public void calculateCGPA() {
        double cumulativeCredit = transcript.getCumulativeCredit();
        double cumulativeGrade = transcript.getCumulativeGrade();
        double result = cumulativeGrade / cumulativeCredit;
        setCgpa((float) result);
    }

    public Transcript getTranscript() {
        return transcript;
    }

    public void setTranscript(Transcript transcript) {
        this.transcript = transcript;
    }


    @Override
    public String toString() {
        return "Student{" + "semester=" + semester + "} " + super.toString() + " advisor=" + advisor;
    }
}
